public/app/features/query/components/QueryGroup.tsx

```diff
@@ -82,8 +82,25 @@ export class QueryGroup extends PureComponent<Props, State> {
       const dsSettings = this.dataSourceSrv.getInstanceSettings(options.dataSource);
       const defaultDataSource = await this.dataSourceSrv.get();
       const datasource = ds.getRef();
-      const queries = options.queries.map((q) => (q.datasource ? q : { ...q, datasource }));
+      let changedQueries = false;
+      const queries = await Promise.all(options.queries.map(async (q) => {
+        if (!q.datasource) {
+          q = { ...q, datasource };
+          changedQueries = true;
+        }
+        if (Object.keys(q).length === 2) {
+          // If query has only refId and datasource, try to set the default query for its datasource
+          q = {...this.newQuery(await this.dataSourceSrv.get(q.datasource), dsSettings), ...q};
+          changedQueries = true;
+        }
+        return q;
+      }));
       this.setState({ queries, dataSource: ds, dsSettings, defaultDataSource });
+      if (changedQueries) {
+        this.onChange({
+          queries
+        });
+      }
```

Comment:
This is basically the same as calling onChange onMount in the query editor, you could maybe refactor out to a function that can also be used from runRequest? Then we can detect and perform this logic before issuing requests as well. And then we don't need to call this.onChange (which should be preserved for use-initiated change)

Reply:
Do we really need to call onChange? If not I think this can be simplified a bit

``` typescript
  const queries = options.queries.map((q) => {
        if (!q.datasource) {
          q = { ...q, datasource };
        }
        if (Object.keys(q).length === 2) {
          q = { ...this.newQuery(ds, dsSettings), ...q };
        }
        return q;
      });
      this.setState({ queries, dataSource: ds, dsSettings, defaultDataSource });
```

Think we may also need to set default query in `[updateQueries](https://github.com/grafana/grafana/blob/mmandrus/fix-testdatadb-default-selection/public/app/features/query/state/updateQueries.ts#L30-L36)` too. Otherwise, this won't work when changing the data source.